ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java

```diff
@@ -118,47 +133,83 @@ public synchronized void trackSerdeOp(
+    final Map<SchemaInfo, Set<String>> keyLoggerSchemas = getLoggerSchemas(
+        kvLoggerNames,
+        IS_KEY
+    );
+
+    final Map<SchemaInfo, Set<String>> valueLoggerSchemas = getLoggerSchemas(
+        kvLoggerNames,
+        IS_VALUE
+    );
+
+    final Set<SchemaInfo> keySchemaInfo = Optional.of(keyLoggerSchemas.keySet())
+        .orElseGet(() -> ImmutableSet.of(EMPTY_SCHEMA));
+    final Set<SchemaInfo> valueSchemaInfo = Optional.of(valueLoggerSchemas.keySet())
+        .orElseGet(() -> ImmutableSet.of(EMPTY_SCHEMA));
+
+    final Set<KeyFormat> keyFormats = keySchemaInfo.stream()
+        .map(s -> s.keyFormat)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .collect(Collectors.toSet());
+
+    if (keyFormats.size() != 1) {
+      final String result = keyFormats.size() == 0 ? "Zero" : "Multiple";
+
+      throw new IllegalStateException(result + " key formats registered for topic."
           + System.lineSeparator()
           + "topic: " + topicName
-          + "loggers: " + keyLoggerNames
+          + "formats: " + keyFormats.stream().map(KeyFormat::getFormat)
+          .sorted().collect(Collectors.toList())
+          + "loggers: " + keyLoggerSchemas.values()
       );
     }
 
-    final Set<String> valueTopicNames = kvLoggerNames.getOrDefault(false, ImmutableSet.of());
-    if (valueTopicNames.size() != 1) {
-      throw new IllegalStateException("Multiple value logger names registered for topic."
+    final Set<ValueFormat> valueFormats = valueSchemaInfo.stream()
+        .map(s -> s.valueFormat)
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .collect(Collectors.toSet());
+
+    if (valueFormats.size() > 1) {
+      throw new IllegalStateException("Multiple value formats registered for topic."
           + System.lineSeparator()
           + "topic: " + topicName
-          + "loggers: " + valueTopicNames
+          + "formats: " + valueFormats.stream().map(ValueFormat::getFormat)
+          .sorted().collect(Collectors.toList())
+          + "loggers: " + valueLoggerSchemas.values()
       );
     }
 
-    final SchemaInfo keyInfo = loggerToSchemas.get(Iterables.getOnlyElement(keyLoggerNames));
-    final SchemaInfo valueInfo = loggerToSchemas.get(Iterables.getOnlyElement(valueTopicNames));
-    if (keyInfo == null || valueInfo == null) {
-      throw new IllegalStateException("Incomplete schema info for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "keyInfo: " + keyInfo
-          + "valueInfo: " + valueInfo
-      );
+    final KeyFormat keyFormat = Iterables.getOnlyElement(keyFormats);
+    final ValueFormat valueFormat = !valueFormats.isEmpty()
+        ? Iterables.getOnlyElement(valueFormats)
+        : NONE_VALUE_FORMAT;
+
+    return new TopicFormatsInfo(keyFormat, valueFormat);
+  }
+
+  private Map<SchemaInfo, Set<String>> getLoggerSchemas(
```

Comment:
Can we add a short javadoc to this? It wasn't clear to me from the name what the method does -- I had to read the code to find out.

Reply:
Done.